### PR TITLE
([B,]C,H,W) image convention

### PR DESCRIPTION
**Pull request by tstreule**
Before, we were quite inconsistent with the dimensions. This PR introduces the (B,C,H,W) and (C,H,W) convention in order to be less error prone.

(B,C,H,W) convention status:

Scripts:
- [x] `arc_intersection.py`
- [x] `train.py`

Functions and files:
- [x] `point_projection.py`
- [x] `keypoint_net_with_io_loss.py`
    - [x] All except IO-loss verified

### CODE REVIEW

File: sips/utils/keypoint_matching.py [hunk 1]
```diff
@@ -60,6 +61,20 @@ def _groupwise_smallest_values_mask(
...
+    groups : torch.Tensor
+        Group identifiers of the entries (usually unraveled indices).
+        Shape: (n_entries,n_groupers)
+    values : torch.Tensor
+        Corresponding values of the entries. Shape: (n_entries,)
+
+    Returns
+    -------
+    torch.Tensor
+        Boolean ask that is True for every groupwise smallest entry.
```
Comment (stygerma) [inline sips/utils/keypoint_matching.py:75]: should probably be
```suggestion
        Boolean mask that is True for every groupwise smallest entry.
```

File: sips/utils/keypoint_matching.py [hunk 2]
```diff
@@ -99,6 +114,80 @@ def _groupwise_smallest_values_mask(
...
+    Parameters
+    ----------
+    tensor : torch.Tensor
+        Tensor with its values. Shape: ([B,]2,H,W) or ([B,]H,W,2)
+    grid_index : torch.LongTensor
+        (Mesh-)grid index for selecting the `tensor` values. See Notes.
+        Shape: Same as `tensor`, ([B,]2,H,W) or ([B,]H,W,2)
+    data_format : {'BCHW', 'BHWC'}, optional
+        Specify the data format of the input and output data. With the default format
+        "NCHW", the data is assumed to be in the order of:
```
Comment (stygerma) [inline sips/utils/keypoint_matching.py:137]: Probably rather
```suggestion
        "BCHW", the data is assumed to be in the order of:
```

File: sips/utils/keypoint_matching.py (hunk 1 above)
Comment (tstreule) [reply sips/utils/keypoint_matching.py:75]: Nice one! Didn't know that you can suggest changes like that

File: sips/utils/plotting.py [hunk 3]
```diff
@@ -34,157 +35,219 @@ def _asarray(arr: npt.ArrayLike) -> npt.NDArray[np.float32]:
...
+    bg_image: torch.Tensor | None = None,
+    *,
+    target_color: str | None = COLOR_TARGET_SONAR,
+    source_color: str | None = COLOR_SOURCE_SONAR,
+    match_color: str | None = COLOR_HIGHLIGHT,
+    s: float | None = None,
 ) -> plt.Axes:
     """
-    Plot keypoint positions in sonar image (uv) space.
+    Plot keypoint ppositions in sonar image (uv) space.
```
Comment (stygerma) [inline sips/utils/plotting.py:51]: was this intentional? if yes, what is `ppositions`?

File: scripts/examples/arc_intersection.py [hunk 4]
```diff
@@ -61,60 +55,55 @@ def main(
...
-    kp1_xyz = uv_to_xyz_batch(kp1_uv, batch.params1, batch.pose1, (width, height))
-    kp2_xyz = uv_to_xyz_batch(kp2_uv, batch.params2, batch.pose2, (width, height))
+    target_xyz = uv_to_xyz_batch(target_uv, batch.params1, batch.pose1, (height, width))
+    source_xyz = uv_to_xyz_batch(source_uv, batch.params2, batch.pose2, (height, width))
 
     # Point projection to other camera
-    # kp2_uv_proj = batch_xyz_to_uv(kp2_xyz, batch.params1, batch.pose1, (width, height))
-    kp2_uv_proj = warp_image_batch(
-        kp2_uv, batch.params2, batch.pose2, batch.params1, batch.pose1, (width, height)
+    # source_uv_proj = batch_xyz_to_uv(kp2_xyz, batch.params1, batch.pose1, (height, width))
```
[the remark this answers is not in the document]
Comment (tstreule) [reply scripts/examples/arc_intersection.py:70]: kk

File: sips/models/keypoint_net_with_io_loss.py [hunk 5]
```diff
@@ -2,33 +2,22 @@
...
 import torch
 import torch.nn.functional as F
-import wandb
 from pytorch_lightning.loggers.wandb import WandbLogger
 from torch import optim
-from torchvision.transforms import ToPILImage
 from typing_extensions import Self
 
-from sips import PSEUDO_INF
+import wandb
```
[the remark this answers is not in the document]
Comment (tstreule) [reply sips/models/keypoint_net_with_io_loss.py:13]: 'cause `black` is a retard. I had a folder called "wandb"

File: sips/models/keypoint_net_with_io_loss.py [hunk 6]
```diff
@@ -57,8 +46,8 @@ def build_descriptor_loss(
     source_des: torch.Tensor,
     target_des: torch.Tensor,
```
[the remark this answers is not in the document]
Comment (tstreule) [reply sips/models/keypoint_net_with_io_loss.py:47]: Should be fine. KP2D did it the same way

File: sips/models/keypoint_net_with_io_loss.py [hunk 7]
```diff
@@ -102,18 +91,20 @@ def build_descriptor_loss(
...
         ref_desc = F.grid_sample(
             source_des[b].unsqueeze(0),
-            source_points[b].unsqueeze(0),
+            source_points[b].unsqueeze(0).nan_to_num(0),
             align_corners=True,
-        ).squeeze()
+        ).squeeze(0)
         tar_desc = F.grid_sample(
             target_des[b].unsqueeze(0),
             # convert nan to large value such that the gradient does not get nan
```
[the remark this answers is not in the document]
Comment (tstreule) [reply sips/models/keypoint_net_with_io_loss.py:103]: I'll just delete the comment

File: sips/models/keypoint_net_with_io_loss.py [hunk 8]
```diff
@@ -180,12 +169,12 @@ def build_descriptor_loss(
...
-        #  Hence for now, we just ignore such losses when they are NaN.
-        if not batch_loss.isnan().any():
-            loss += batch_loss / B
+        if batch_loss.isnan():
+            # NaNs would corrupt all model parameters on tensor.backward()
+            raise RuntimeError(f"Descriptor loss for batch {b} is NaN.")
+        loss += batch_loss / B
 
-    return loss, recall
+    return loss, recall.detach()
```
[the remark this answers is not in the document]
Comment (tstreule) [reply sips/models/keypoint_net_with_io_loss.py:177]: The recall is not used for the gradient. If you look closely, the code before was like that:

```python
recall += float(1.0 / B) * (float(correct_matches.float().sum()) / float(ref_desc.size(1)))
```

This implicitly also "detached" the recall by converting it into Python a float object

File: sips/utils/plotting.py [hunk 9]
```diff
@@ -34,157 +35,219 @@ def _asarray(arr: npt.ArrayLike) -> npt.NDArray[np.float32]:
...
     Parameters
     ----------
-    keypoints_xyz : Sequence[npt.ArrayLike]
-        Keypoints of shape (..., n_elevations, 3).
-    camera_pos : Sequence[npt.ArrayLike | None] | None, optional
-        Camera position for the (set of) keypoints, by default None
-    colors : str | Sequence[str] | None, optional
-        Colors of the keypoints, by default None
+    projection_arcs : Sequence[torch.Tensor]
+        Projection arcs to be plotted. Shape: (N_ELEVATIONS,3,...)
```
Comment (stygerma) [inline sips/utils/plotting.py:157]: Emphasize that this is a Sequence
```suggestion
        Projection arcs to be plotted. Shape: [(N_ELEVATIONS,3,...), ...]
```

File: scripts/examples/arc_intersection.py [hunk 10]
```diff
@@ -61,60 +55,55 @@ def main(
...
-        # fmt: off
-        kp1_xyz_b = kp1_xyz_b.reshape(W, H, D, 3)[::4, ::8].reshape(-1, D, 3)
-        kp2_xyz_b = kp2_xyz_b.reshape(W, H, D, 3)[::4, ::8].reshape(-1, D, 3)
-        kp2_xyz_filtered = kp2_xyz_filtered.reshape(W, H, D, 3)[::4, ::4].reshape(-1, D, 3)
-        # fmt: on
+        # Find source points that are visible in target image
+        source_invisible = source_uv_proj.isnan().any(dim=2, keepdim=True)
+        source_xyz_visible = source_xyz.masked_fill(source_invisible, torch.nan)
+
         plot_arcs_3d(
```
Comment (stygerma) [inline scripts/examples/arc_intersection.py:93]: Does not work for me. Looking into it

File: sips/models/keypoint_net_with_io_loss.py [hunk 11]
```diff
@@ -335,122 +331,157 @@ def _get_loss_recall(
...
+                [H, W],
+                self.cell,
+                matches=[
+                    target_uv_match[b, matching_mask[b]],
+                    source_uv_warp_match[b, matching_mask[b]],
+                ],
+                bg_image=batch[b].image1,
+            )
+            ax.set_title("Debug: Keypoint/Localization loss")
+            # plt.show(block=False)
```
[the remark this answers is not in the document]
Comment (tstreule) [reply sips/models/keypoint_net_with_io_loss.py:399]: yessir :+1:

File: sips/models/keypoint_net_with_io_loss.py [hunk 12]
```diff
@@ -335,122 +331,157 @@ def _get_loss_recall(
...
-        ).view(B, HC * WC)
+        )
         mse_loss = F.mse_loss(
-            target_score_resampled[mask], source_score.view(B, HC * WC)[mask]
+            target_score_resampled.squeeze(1)[mse_mask],
+            source_score.squeeze(1)[mse_mask],
         )
 
-        loss_2d += self.score_loss_weight * (usp_loss.mean() + 2 * mse_loss)
+        loss_2d += self.score_loss_weight * (usp_loss.mean() + mse_loss)
```
[the remark this answers is not in the document]
Comment (tstreule) [reply sips/models/keypoint_net_with_io_loss.py:459]: That was a bug previously. Before, I halved `usp_loss` and doubled the `mse_loss` and thus had a factor equalling 4 instead of 2.

File: sips/networks/keypoint_net.py [hunk 13]
```diff
@@ -102,10 +100,12 @@ def __init__(
             torch.nn.Conv2d(c4, c5, kernel_size=3, stride=1, padding=1, bias=False),
             torch.nn.BatchNorm2d(c4, momentum=self.bn_momentum),
         )
-        self.convFb = torch.nn.Sequential(
-            torch.nn.Conv2d(c5, d1, kernel_size=3, stride=1, padding=1, bias=False),
-            torch.nn.BatchNorm2d(d1, momentum=self.bn_momentum),
-        )
+        if self.do_upsample:
```
[the remark this answers is not in the document]
Comment (tstreule) [reply sips/networks/keypoint_net.py:103]: It is a bug fix. Looking into our base_config.py file, we before had `strategy = "ddp_find_unused_parameters_true"` and now it is `strategy = "ddp"`. Without this `if`-case the latter would result in an error because for example `self.convFb` would be an unused parameter if we don't upsample and the `loss.backward()` step would raise an error because it couldn't find a gradient for backpropagation.

File: sips/networks/keypoint_net.py [hunk 14]
```diff
@@ -185,31 +180,24 @@ def forward(self, x: torch.Tensor) -> _forward_return_type:
...
-                Hc,
-                Wc,
-                dtype=center_shift.dtype,
-                device=center_shift.device,
-                ones=False,
-                normalized=False,
-            ).mul(self.cell)
-            + step
-        )
+        step = self.cell / 2.0
```
[the remark this answers is not in the document]
Comment (tstreule) [reply sips/networks/keypoint_net.py:187]: That's something I briefly discussed with you. I don't see how the previous formulation made sense. Assume the cell size is 8, then why should the step/offset equal 3.5? The center of the cell is at 4.

File: sips/utils/plotting.py (hunk 3 above)
Comment (tstreule) [reply sips/utils/plotting.py:51]: Of course not 😄

File: sips/utils/keypoint_matching.py [hunk 15]
```diff
@@ -388,106 +505,109 @@ def match_keypoints_2d(
...
+    _sanity_check_coords(target_uv, cell_size, H, W)
+
+    # Create sliding window view
+    WS = window_size
+    PAD = (window_size - 1) // 2
+    target_uv_strided = (
+        F.pad(target_uv, 4 * [PAD], mode="constant", value=0)  # (B,2,H+2PAD,W+2PAD)
+        .unfold(2, WS, 1)
+        .unfold(3, WS, 1)
+    )  # (B,2,H,W,WS,WS)
```
[the remark this answers is not in the document]
Comment (tstreule) [reply sips/utils/keypoint_matching.py:542]: Yes.

File: sips/utils/keypoint_matching.py [hunk 16]
```diff
@@ -388,106 +505,109 @@ def match_keypoints_2d(
...
+
+    # Permute and flatten dimensions for nicer data handling
+    target_uv_strided = target_uv_strided.movedim(1, 3)  # (B,H,W,2,WS,WS)
```
[the remark this answers is not in the document]
Comment (tstreule) [reply sips/utils/keypoint_matching.py:545]: Yes.

File: sips/utils/plotting.py (hunk 9 above)
Comment (tstreule) [reply sips/utils/plotting.py:157]: Can do that

File: sips/utils/plotting.py [hunk 17]
```diff
@@ -34,157 +35,219 @@ def _asarray(arr: npt.ArrayLike) -> npt.NDArray[np.float32]:
...
+    camera_positions : Sequence[torch.Tensor  |  None]
+        Camera positions for the projection arcs. Required to plot the camera origin
+        with a dot. Shape: (3,)
```
[the remark this answers is not in the document]
Comment (tstreule) [reply sips/utils/plotting.py:160]: Yup

**Review by stygerma — COMMENTED** [review]
Very well done

File: sips/models/keypoint_net_with_io_loss.py [hunk 18]
```diff
@@ -678,11 +680,23 @@ def _shared_eval(self, batch: SonarBatch, batch_idx: int, prefix: str) -> None:
 
             # In debugging, you can get until here...
             if self._debug:
-                _, axs = plt.subplots(2, 2)
+                fig, axs = plt.subplots(2, 3, figsize=(15, 10))
```
Comment (stygerma) [inline sips/models/keypoint_net_with_io_loss.py:683]: maybe include some info about what each subplot depicts

Comment (tstreule) [reply sips/models/keypoint_net_with_io_loss.py:683]: Good idea

**Review by stygerma — APPROVED** [review]
